### PR TITLE
SQLViewData is not read only when filtering

### LINKED ISSUE #152: You should be able to edit your data in this mode
JIRA ref: https://cartodb.atlassian.net/browse/CDB-238

### PR DESCRIPTION
**Pull request by alonsogarciapablo**
Fixes #152. Basically: you will now be able to edit cells while filtering, but not when a sql query has been applied, and active filters are updated accordingly :)

### CODE REVIEW

File: lib/assets/javascripts/cartodb/models/sqlview.js [hunk 1]
```diff
@@ -99,10 +95,10 @@ cdb.admin.SQLViewData = cdb.admin.CartoDBTableData.extend({
     });
   },
 
+  // TODO: Is this being used?
   filterColumn: function(columnName, tableName, filter) {
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb/models/sqlview.js]: I found two references to this method:

- [This one (it seems we're not using that dropdown anymore)](https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/public_table/header_dropdown_public.js#L29)
- and [this other one (commented out)](https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/table/filter_column_dialog.js#L60)

@xavijam do you think I can remove it?

File: lib/assets/javascripts/cartodb/models/sqlview.js [hunk 2]
```diff
@@ -77,9 +72,10 @@ cdb.admin.SQLViewData = cdb.admin.CartoDBTableData.extend({
   },
 
   isReadOnly: function() {
-    return this.readOnly;
+    return this.sqlSource() !== 'filters';
```
Comment (javisantana) [inline lib/assets/javascripts/cartodb/models/sqlview.js]: umh, I don't think this is a good idea, we set readonly in different cases (sync tables for example)

**Comment by saleiva** [comment]
THIS IS FUCKING GOOD

**Comment by xavijam** [comment]
@javisantana, please, could you take care of this PR?

File: lib/assets/javascripts/cartodb/models/sqlview.js (hunk 2 above)
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb/models/sqlview.js]: That's happening [at the `cdb.admin.CartoDBTableData` level](https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/models/table.js#L174). I think we're not setting readonly in `cdb.admin.SQLViewData` from the outside.

So, with this change:

- Synced tables will always be read only.
- Non synced tables will be read only if:
  - User doesn't have permissions to edit the table (that's when we're setting readonly at the table level).
  - A SQL query is being applied (a custom SQL query, not a filter).

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3960/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4088/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4128/))

**Comment by alonsogarciapablo** [comment]
@javisantana This is ready for review again. There was some hidden work here: filters were not being re-rendered correctly when the data was saved because:
- `data:saved` was not being triggered correctly.
- histograms were not prepared to respond to "expansion" of lower and upper values.

Thx to @javierarce for helping me out with the histogram issue!

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4184/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4185/))

File: lib/assets/test/spec/cartodb/models/filters.spec.js [hunk 3]
```diff
@@ -1,5 +1,13 @@
 describe('modules.Filters', function() {
 
+  var clone = function(obj) {
```
Comment (javisantana) [inline lib/assets/test/spec/cartodb/models/filters.spec.js]: why not _.clone?

**Comment by alonsogarciapablo** [comment]
Hey @javierarce, can you please take a look at this PR? Thanks!

File: lib/assets/test/spec/cartodb/models/filters.spec.js (hunk 3 above)
Comment (alonsogarciapablo) [inline lib/assets/test/spec/cartodb/models/filters.spec.js]: Fixed! Thanks!

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4189/))

**Comment by javierarce** [comment]
@alonsogarciapablo: code looks good to me :+1:

File: lib/assets/javascripts/cartodb/models/sqlview.js [hunk 4]
```diff
@@ -77,32 +72,7 @@ cdb.admin.SQLViewData = cdb.admin.CartoDBTableData.extend({
...
-      table: tableName,
-      filter: filter
-    });
-  },
-
-  filterColumn: function(columnName, tableName, filter) {
-    // filter table by column
-    this.setSQL(this.filterColumnSQL(columnName, tableName, filter));
-    this.readOnly = false;
+    return this.sqlSource() !== 'filters';
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb/models/sqlview.js]: Another thing that could check here to determine if a row is editable:

`return !!this.getSQL().indexOf("SELECT * FROM " + this.table.get('id'))`

This way, cells will be editable when filtering or when a SQL query is applied an all the columns are being selected.

cc/ @javisantana

**Comment by saleiva** [comment]
Should it work on the Map page too? I'm not able to edit my data in the map page (dataset view) when filtering on a category column...

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/4221/))

**Comment by alonsogarciapablo** [comment]
Yes, it should work:

filtering

**Comment by saleiva** [comment]
weird, now it works :) Thanks!

**Pull request merged**